File: some_javascript/divide-conquer-sorting-searching-random-algs/assignment2/merge-sort.js
function mergeSort(array, arrayLength) {
  if (arrayLength === 1) {
      return array;
  }

  var half = Math.floor(arrayLength / 2);
  var firstHalf = array.slice(0, half);
  var secondHalf = array.slice(half);

  var firstSorted = mergeSort(firstHalf, firstHalf.length)
  var secondSorted = mergeSort(secondHalf, secondHalf.length);
  var mergeHalfs = merge(firstSorted, secondSorted, arrayLength);

  return mergeHalfs
}

function merge(array1, array2, sumLength) {
  var result = []; 
  for (var k = 0; k < sumLength; k++) {
    if (array1.length === 0 || array2.length === 0) {
      result = result.concat(array1);
      result = result.concat(array2);
      break;
    }
    if (array1[0] < array2[0]) {
      result.push(array1.shift());
    } else {
      result.push(array2.shift());
    }
  }
  return result;
}

module.exports = mergeSort;

File: divide-conquer-randomized-algs/assignment2/merge_sort.py
def merge_sort(array, arraylen):
    if arraylen == 1:
        return array

    half = arraylen // 2
    first_half = array[:half]
    second_half = array[half:]

    first_sorted = merge_sort(first_half, len(first_half))
    second_sorted = merge_sort(second_half, len(second_half))
    merge_halfs = merge(first_sorted, second_sorted, arraylen)

    return merge_halfs


def merge(array1, array2, sumlen):
    result = []

    for _ in range(sumlen):
        if not array1 or not array2:
            result.extend(array1)
            result.extend(array2)
            break
        if array1[0] < array2[0]:
            result.append(array1.pop(0))
        else:
            result.append(array2.pop(0))

    return result
